global/scripts/debug.shlib

```diff
@@ -0,0 +1,272 @@
+
+debug_disable_color()
+{
+    HCP_DEBUG_COLOR=FALSE
+}
+
+
+# Description:
+#
+#   Function for printing red debug output.
+#
+# Input:
+#
+#   — text to be printed
+
+debug_red_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[31m $* \033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+# Description:
+#
+#   Function for printing green debug output.
+#
+# Input:
+#
+#   — text to be printed
+
+debug_green_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[32m $* \033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+# Description:
+#
+#   Function for printing blue debug output.
+#
+# Input:
+#
+#   — text to be printed
+
+debug_blue_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[34m $* \033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+debug_on_error()
+{    
+    if [ ! $HCP_DEBUG_USETRAP == 'OFF' ]; then        
+
+        local me=`basename "$0"`
+        local line
+        local startreport = ""
+        local endreport   = ""
+
+        # --- try to identify command if $command is not set
+
+        if [ -z "$command" ]; then
+            line=($(eval echo "$2"))
+            command=${line[0]}
+        fi
+
+        # --- output a regular HCP error log
+
+        log_Err "$command command failed with return code: $3"
+
+        # --- are we printing in color
+
+        if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+           startreport="\033[33m"
+           endreport="\033[0m"
+        fi
+        
+        # --- output a detailed report if usetrap is not turned off    
+        echo                                                               
+        echo -e "${startreport}##### ERROR: Execution stopped prematurely #####"
+        echo "---> script: $me"                                            
+        echo "---> stopped at line: $1"                                    
+        echo "---> command: $command"                                      
+        echo "---> expanded: $(eval echo "$2")"                            
+        echo "---> exit status: $3"                                        
+        echo -e "##### -------------------------------- #####${endreport}" 
+        echo                                                               
+        
+        # --- check if ok to exit        
+        if [ ! $HCP_DEBUG_EXIT_ON_ERROR = 'OFF' ]; then
+            exit $3
+        fi
+    fi
+}
+
+
+
+# ============ VERBOSE functionality ==============
+
+
+# Description:
+#
+#   Function to turn on verbose reporting
+
+debug_enable_verbose()
+{
+    HCP_VERBOSE=TRUE
+}
+
+# Description:
+#
+#   Function to turn off output coloring
+
+debug_disable_verbose()
+{
+    HCP_VERBOSE=FALSE
+}
+
+
+# Description:
+#
+#   Function for printing red verbose output.
+#
+# Input:
+#
+#   — text to be printed
+
+verbose_red_echo()
+{   
+    if [ "$HCP_VERBOSE" == "TRUE" ] ; then
+        if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
```

Comment:
After testing `HCP_VERBOSE`, these could call debug_red_echo, etc, instead of repeating their code.

Reply:
Good point. Done.

Reply:
Good point. Done.